src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs

```diff
@@ -0,0 +1,405 @@
+            config.stderr.clone(),
+            config.terminal,
+        );
+
+        Ok(Self {
+            pid,
+            container_id,
+            config,
+            inner: Arc::new(RwLock::new(ContainerInner::new(
+                agent.clone(),
+                init_process,
+                logger.clone(),
+            ))),
+            agent,
+            resource_manager,
+            logger,
+        })
+    }
+
+    pub async fn create(&self, mut spec: oci::Spec) -> Result<()> {
+        // process oci spec
+        let mut inner = self.inner.write().await;
+        let config = &self.config;
+        amend_spec(&mut spec).context("load spec")?;
+
+        // handler rootfs
+        let rootfs = self
+            .resource_manager
+            .handler_rootfs(&config.container_id, &config.bundle, &config.rootfs_mounts)
+            .await
+            .context("handler rootfs")?;
+
+        // update rootfs
+        match spec.root.as_mut() {
+            Some(spec) => {
+                spec.path = rootfs
+                    .get_guest_rootfs_path()
+                    .await
+                    .context("get guest rootfs path")?
+            }
+            None => return Err(anyhow!("spec miss root field")),
+        };
+        inner.rootfs.push(rootfs);
+
+        // handler volumes
+        let volumes = self
+            .resource_manager
+            .handler_volumes(&config.container_id, &spec.mounts)
+            .await
+            .context("handler volumes")?;
+        let mut oci_mounts = vec![];
+        let mut storages = vec![];
+        for v in volumes {
+            let mut volume_mounts = v.get_volume_mount().context("get volume mount")?;
+            if !volume_mounts.is_empty() {
+                oci_mounts.append(&mut volume_mounts);
+            }
+
+            let mut s = v.get_storage().context("get storage")?;
+            if !s.is_empty() {
+                storages.append(&mut s);
+            }
+            inner.volumes.push(v);
+        }
+        spec.mounts = oci_mounts;
+
+        // TODO: handler device
+
+        // update cgroups
+        self.resource_manager
+            .update_cgroups(
+                &config.container_id,
+                spec.linux
+                    .as_ref()
+                    .map(|linux| linux.resources.as_ref())
+                    .flatten(),
+            )
+            .await?;
+
+        // create container
+        let r = agent::CreateContainerRequest {
+            process_id: agent::ContainerProcessID::new(&config.container_id, ""),
+            string_user: None,
+            devices: vec![],
+            storages,
+            oci: Some(spec),
+            guest_hooks: None,
+            sandbox_pidns: false,
+            rootfs_mounts: vec![],
+        };
+
+        self.agent
+            .create_container(r)
+            .await
+            .context("agent create container")?;
+        self.resource_manager.dump().await;
```

Comment:
Is this line for debugging?

Reply:
Thanks for your review, it is trying to print resource info part of debugging, maybe remove this when supporting runtime metrics.